src/dict.c

```diff
@@ -61,20 +61,31 @@ typedef struct {
 
 /* -------------------------- private prototypes ---------------------------- */
 
-static void _dictExpandIfNeeded(dict *d);
+static int _dictExpandIfNeeded(dict *d);
 static void _dictShrinkIfNeeded(dict *d);
 static void _dictRehashStepIfNeeded(dict *d, uint64_t visitedIdx);
 static signed char _dictNextExp(unsigned long size);
 static int _dictInit(dict *d, dictType *type);
 static dictEntry *dictGetNext(const dictEntry *de);
-static dictEntry **dictGetNextRef(dictEntry *de);
+static dictEntryLink dictGetNextLink(dictEntry *de);
 static void dictSetNext(dictEntry *de, dictEntry *next);
-static int dictDefaultCompare(dict *d, const void *key1, const void *key2);
+static int dictDefaultCompare(dictCmpCache *cache, const void *key1, const void *key2);
+static dictEntryLink dictFindLinkInternal(dict *d, const void *key, dictEntryLink *bucket);
+dictEntryLink dictFindLinkForInsert(dict *d, const void *key, dictEntry **existing);
+static dictEntry *dictInsertKeyAtLink(dict *d, void *key, dictEntryLink link);
+
+/* -------------------------- unused  --------------------------- */
+void dictSetSignedIntegerVal(dictEntry *de, int64_t val);
+int64_t dictGetSignedIntegerVal(const dictEntry *de);
+double dictIncrDoubleVal(dictEntry *de, double val);
+void *dictEntryMetadata(dictEntry *de);
+int64_t dictIncrSignedIntegerVal(dictEntry *de, int64_t val);
 
 /* -------------------------- misc inline functions -------------------------------- */
 
-typedef int (*keyCmpFunc)(dict *d, const void *key1, const void *key2);
-static inline keyCmpFunc dictGetKeyCmpFunc(dict *d) {
+typedef int (*keyCmpFunc)(dictCmpCache *cache, const void *key1, const void *key2);
+static inline keyCmpFunc dictGetCmpFuncAndResetCache(dict *d, dictCmpCache *cache) {
+    memset(cache, 0, sizeof(dictCmpCache));
```

Comment:
```
dictCmpCache cmpCache = {0};
keyCmpFunc cmpFunc = dictGetCmpFunc(d);
```
It feels more readable to initialize dictCmpCache like this.
and change dictGetCmpFuncAndResetCache() to dictGetCmpFunc()